tooling/templatize/pkg/ev2/pipeline.go

```diff
@@ -10,73 +11,109 @@ import (
 	if err != nil {
-		return "", err
+		return nil, fmt.Errorf("failed to read referenced files of pipeline %s: %w", originalPipeline.PipelineFilePath(), err)
+	}
+
+	// precompile the pipeline and referenced files
+	processedPipeline, processedFiles, err := processPipelineForEV2(originalPipeline, referencedFiles, vars)
+	if err != nil {
+		return nil, err
+	}
+
+	// store the processed files to disk relative to the pipeline directory
+	_, restoreDir, err := processedPipeline.EnterPipelineDir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to enter pipeline directory: %w", err)
+	}
+	defer restoreDir()
+	for filePath, content := range processedFiles {
+		err := os.WriteFile(filePath, content, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to write precompiled file %q: %w", filePath, err)
+		}
 	}
 
-	return filepath.Join(pipelineDir, newPipelineFileName), nil
+	return processedPipeline, nil
 }
 
-func processPipelineForEV2(p *pipeline.Pipeline, vars config.Variables) error {
-	_, scopeBoundVars := EV2Mapping(vars, []string{})
+func readReferencedPipelineFiles(p *pipeline.Pipeline) (map[string][]byte, error) {
+	// switch to pipeline directory to ensure relative paths are resolvable
+	_, restoreDir, err := p.EnterPipelineDir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to enter pipeline directory: %w", err)
+	}
+	defer restoreDir()
+
+	referencedFiles := make(map[string][]byte)
 	for _, rg := range p.ResourceGroups {
 		for _, step := range rg.Steps {
 			if step.Parameters != "" {
-				newParameterFilePath, err := precompileFileAndStore(step.Parameters, scopeBoundVars)
+				paramFileContent, err := os.ReadFile(step.Parameters)
 				if err != nil {
-					return err
+					return nil, fmt.Errorf("failed to read parameter file %q: %w", step.Parameters, err)
 				}
-				step.Parameters = newParameterFilePath
+				referencedFiles[step.Parameters] = paramFileContent
 			}
 		}
 	}
-	return nil
+	return referencedFiles, nil
 }
 
-func precompileFileAndStore(filePath string, vars map[string]interface{}) (string, error) {
-	preprocessedBytes, err := config.PreprocessFile(filePath, vars)
+func processPipelineForEV2(p *pipeline.Pipeline, referencedFiles map[string][]byte, vars config.Variables) (*pipeline.Pipeline, map[string][]byte, error) {
+	processingPipeline, err := p.DeepCopy(buildPrefixedFilePath(p.PipelineFilePath(), precompiledPrefix))
 	if err != nil {
-		return "", err
+		return nil, nil, err
 	}
-	newFilePath := buildPrefixedFilePath(filePath, "ev2-precompiled-")
-	err = os.WriteFile(newFilePath, preprocessedBytes, 0644)
-	if err != nil {
-		return "", err
+	processedFiles := make(map[string][]byte)
+	_, scopeBoundVars := EV2Mapping(vars, []string{})
+	for _, rg := range processingPipeline.ResourceGroups {
+		for _, step := range rg.Steps {
+			// preprocess the parameters file with scopebinding variables
+			if step.Parameters != "" {
+				paramFileContent, ok := referencedFiles[step.Parameters]
+				if !ok {
```

Comment:
Is this check necessary?  isn't line 79 covering this?

Reply:
79 is in a different function altogether right?